### PR TITLE
prevent multiple demand in websocket-core

### LINKED ISSUE #9965: prevent multiple websocket frames from being demanded in Jetty-12
**Jetty version(s)**
12

**Description**
After PR #9945 you can no longer demand multiple frames at once. 
We should remove the internal counter tracking demand and throw ISE or ReadPending exception if multiple frames are demanded.

ie. this code should be throwing ISE but instead it is just incrementing the internal demand counter.
```java
for (long i = 0; i < n; i++)
           demand();
```

### PR DESCRIPTION
**Pull request by lachlan-roberts**
closes #9965

### CODE REVIEW

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java [hunk 1]
```diff
@@ -61,7 +61,7 @@ public class WebSocketConnection extends AbstractConnection implements Connectio
     private final WebSocketCoreSession coreSession;
     private final Flusher flusher;
     private final Random random;
-    private long demand;
+    private Boolean demand = false;
```
Comment (gregw) [inline jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java]: I know I often use a Boolean to create a tri-state.... but it is often a mistake.   Would an enum be clearer here?   If you keep the Boolean, then at least add a comment what the null value means.

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java (hunk 1 above)
Comment (sbordet) [reply jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java]: I would also prefer an `enum`.

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/DemandingFlusher.java [hunk 2]
```diff
@@ -139,8 +141,8 @@ public void failFlusher(Throwable t)
      */
     public void emitFrame(Frame frame, Callback callback)
     {
-        if (_demand.decrementAndGet() < 0)
-            throw new IllegalStateException("Negative Demand");
+        if (!_demand.compareAndSet(true, false))
+            throw new IllegalStateException("Demand Already Fulfilled");
```
Comment (sbordet) [inline jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/DemandingFlusher.java]: De-capitalize 2nd and 3rd word.

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/DemandingFlusher.java [hunk 3]
```diff
@@ -41,7 +42,7 @@ public abstract class DemandingFlusher extends IteratingCallback implements Dema
     private static final Throwable SENTINEL_CLOSE_EXCEPTION = new StaticException("Closed");
 
     private final IncomingFrames _emitFrame;
-    private final AtomicLong _demand = new AtomicLong();
+    private final AtomicBoolean _demand = new AtomicBoolean();
```
Comment (sbordet) [inline jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/DemandingFlusher.java:45]: Why a tri-state that was necessary before is not needed now?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/DemandingFlusher.java (hunk 3 above)
Comment (lachlan-roberts) [reply jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/DemandingFlusher.java:45]: Well it was a `long` before, so it was like this before just in a different way.

demand>0 was for demanding
demand==0 was for no demand
demand<0 was for cancelled demand

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java]: done

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java [hunk 4]
```diff
@@ -351,10 +358,12 @@ public void demand()
             if (LOG.isDebugEnabled())
                 LOG.debug("demand {} d={} fp={} {}", demand, fillingAndParsing, networkBuffer, this);
 
-            if (demand < 0)
-                return;
-
-            demand = MathUtils.cappedAdd(demand, 1);
+            if (demand != null)
```
Comment (sbordet) [inline jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java]: `demand != CANCELLED` instead?

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java [hunk 5]
```diff
@@ -397,14 +415,13 @@ public boolean meetDemand()
...
-            if (demand == 0)
+            if (demand == DemandState.NOT_DEMANDING)
                 throw new IllegalStateException();
             if (!fillingAndParsing)
                 throw new IllegalStateException();
 
-            if (demand > 0)
-                demand--;
-
+            if (demand != null)
```
Comment (sbordet) [inline jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java]: `demand != CANCELLED` instead?

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java [hunk 6]
```diff
@@ -415,7 +432,7 @@ public void cancelDemand()
         {
             if (LOG.isDebugEnabled())
                 LOG.debug("cancelDemand d={} fp={} {} {}", demand, fillingAndParsing, networkBuffer, this);
-            demand = -1;
+            demand = null;
```
Comment (sbordet) [inline jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java]: `demand = CANCELLED` instead?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java [hunk 7]
```diff
@@ -54,14 +54,21 @@ public class WebSocketConnection extends AbstractConnection implements Connectio
      */
     private static final int MIN_BUFFER_SIZE = Generator.MAX_HEADER_LENGTH;
 
+    private enum DemandState
+    {
+        DEMANDING,
+        NOT_DEMANDING,
```
Comment (gregw) [inline jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java:60]: `IDLE` would also be OK.

**Review by gregw — APPROVED** [review]

**Review by sbordet — APPROVED** [review]

**Pull request merged**